cmd/api.go

```diff
@@ -0,0 +1,649 @@
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"runtime"
+	"strings"
+	"sync"
+
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/types"
+	tmclient "github.com/cosmos/cosmos-sdk/x/ibc/light-clients/07-tendermint/types"
+	"github.com/cosmos/relayer/helpers"
+	"github.com/cosmos/relayer/relayer"
+	"github.com/gorilla/mux"
+	"github.com/spf13/cobra"
+)
+
+// ServicesManager represents the manager of the various services the relayer is running
+type ServicesManager struct {
+	Services map[string]*Service
+
+	sync.Mutex
+}
+
+// NewServicesManager returns a new instance of a services manager
+func NewServicesManager() *ServicesManager {
+	return &ServicesManager{Services: make(map[string]*Service)}
+}
+
+// Service represents a relayer listen service
+// TODO: sync services to disk so that they can survive restart
+type Service struct {
+	Name   string `json:"name"`
+	Path   string `json:"path"`
+	Src    string `json:"src"`
+	SrcKey string `json:"src-key"`
+	Dst    string `json:"dst"`
+	DstKey string `json:"dst-key"`
+
+	doneFunc func()
+}
+
+// NewService returns a new instance of Service
+func NewService(name, path string, src, dst *relayer.Chain, doneFunc func()) *Service {
+	return &Service{name, path, src.ChainID, src.Key, dst.ChainID, dst.Key, doneFunc}
+}
+
+const (
+	version    = "version"
+	cfg        = "config"
+	chains     = "chains"
+	paths      = "paths"
+	keys       = "keys"
+	nameArg    = "{name}"
+	status     = "status"
+	header     = "header"
+	height     = "height"
+	query      = "query"
+	lightPath  = "light"
+	chainIDArg = "{chain-id}"
+)
+
+// Middleware calls initConfig for every request
+func Middleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_ = initConfig(rootCmd)
+		h.ServeHTTP(w, r)
+	})
+}
+
+func getAPICmd() *cobra.Command {
+	apiCmd := &cobra.Command{
+		Use: "api",
+		// Aliases: []string{""},
+		Short: "Start the relayer API",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			r := mux.NewRouter()
+
+			r.Use(Middleware)
+
+			sm := NewServicesManager()
+			// NOTE: there is no hardening of this API. It is meant to be run in a secure environment and
+			// accessed via private networking
+
+			// VERSION
+			// version get
+			r.HandleFunc(fmt.Sprintf("/%s", version), VersionHandler).Methods("GET")
```

Comment:
these are really hard to read, what's the need for `fmt.Sprintf`? I think the value of having the endpoints as constants is less than using a plain string